SCRD-4675 openstack-ardana Jenkins pipeline job

stefannica opened the pull request:
Implements: https://jira.suse.de/browse/SCRD-4675
Test job preview: 
 - regular UI: https://ci.suse.de/job/openstack-ardana-pipeline/
 - BlueOcean: https://ci.suse.de/blue/organizations/jenkins/openstack-ardana-pipeline/activity

Re-organizes the openstack-ardana Jenkins job and its downstream
jobs as Jenkins pipeline jobs, bringing along the following
improvements:

 * aside from the job parameters, which are still coverd by the
 JJB descriptor, the job implementation resides completely in a
 Jenkinsfile. This significantly simplifies self-testing as well
 as testing in-progress changes, as one only need to point the job
 run to a different automation repository fork and branch.

 * most of the pipeline stages are implemented by external bash
 scripts or ansible playbooks, which means the pipeline definition
 itself will not be subject to frequent modifications

 * the pipeline structure as well as the status of individual stages
 can be monitored more effectively in the BlueOcean Jenkins UI

 * jobs that fail due to environmental issues can be replayed starting
 with the failed stage

 * stages that don't depend on each-other are run in parallel, to reduce
 execution time

 * for parallel stages, the `fast-fail` pipeline feature is enabled and
 will cancel all constituent stages in a parallel block if even one of those
 stages fails. NOTE: this will be expanded further, to cover the entire set of
 independent validation tests being executed for a single Gerrit patch

 * the pipeline "lockable resources" feature is used to throttle the number
 of parallel builds for an individual stage or for an entire job

 * reusable pipeline stages are implemented as parameterized downstream jobs,
 which can also be manually (re)built independently, as well as re-played
 from the parent pipeline job, if needed. The following downstream jobs are
 available:

   - openstack-ardana-testbuild-gerrit: this job creates test IBS packages
   corresponding to supplied Gerrit patches

   - openstack-ardana-vcloud: builds a virtual cloud environment
   corresponding to a CLM input model, which itself can either be predefined
   or generated using the input model generator feature

   - openstack-ardana-bootstrap: sets up a deployer node (adds/updates media
   and repositories, SLES and RHEL artifacts, installs/updates the ardana
   pattern, etc). The result is a deployer node that can be used to deploy
   any cloud input model. Currently, the deployer node is used by a single
   Ardana deployment, but having this stage defined as a standalone
   downstream job enables us to snapshot it and clone it to be used by several
   test scenario at once

   - openstack-ardana-tempest: sets up tempest and runs tempest test cases on
   a predeployed cloud

 * allows using multiple agent nodes (the current openstack-ardana job
 dos not handle that well) and addresses the problem or running dependent
 stages that need to share large amounts of information (e.g. generated
 input models, updated ansible repositories) on different agents. Every
 downstream job has the option of reusing a node/workspace.

This PR also does the following related changes:
 * changes the ECP resource management associated with running jobs. The
 new `clm_env` parameter is used to uniquely identify a virtual
 or hardware/QA cloud environment, as well as the workspace that a pipeline
 job run uses. The IBS and Gerrit triggered CI pipeline job will use a limit
ed
 number of lockable resources representing virtual slots, instead of the
 non-centralized and non-intuitive build pool logic currently implemented
 by the openstack-ardana-heat job.

 * the ansible inventory management and ansible playbooks have been updated
 to run on virtual environments

JanZerebecki commented on jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile:
```diff
@@ -0,0 +1,214 @@
+      }
+    }
+
+    stage('generate heat') {
```
This is just a random location in the pull changes to group this discussion.
### looser coupling with Jenkins
I'd like for most of what is done in Jenkins jobs to work outside of a Jenkins server (say on a development machine). For normal Jobs that is a matter of just using one JJB builder that calls a script file instead of inlining scripts in the JJB job definitions. https://github.com/SUSE-Cloud/automation/pull/2671/ makes a change in that direction.

For Jenkins pipelines this is not as easy as Jenkins needs to know at least the steps, their order and which ones can be run in parallel. Duplicating this information in a script file that can be executed and a Jenkinsfile has the downside of duplication, which I'd like to avoid.

Maybe there is already a way to run a Jenkins pipeline without a Jenkins server installation?

### Idea how to solve this:
Lets say we have cloud.py (could be acloud, mkcloud, an ansible playbook/extension, a yml file with some python code to run it) which does what the Jenkins job should do.

We implement cloud.py --generate-jenkins which internally uses the same information it would when running by itself to generate the Jenkinsfile. This might be internally represented as a list of steps with a special parallel step that contains a list of steps to execute in parallel. This internal representation can then be used to run them all. And to generate a CLI to call them individually like testcloud.py --step step1. The CLI for an individual step is then used in the generated Jenkins file to run the code for the individual steps.

To call the Jenkins pipeline we have a normal job that checks out the repo where testcloud.py exists, then generates the Jenkinsfile with it, then create the Jenkins subjob to run the pipeline file that was just generated.

JanZerebecki replied:
Just posting this here so I'll find this again. Example what it might look like in Ansible:
```
- import_playbook: "{{ item.playbook }}"
  when: "{{ item.when }}"
  loop:
  - { playbook: generate-input-model.yml, when: True }
  - { playbook: rebuild-deployer-vm.yml, when: True }
  - { playbook: bootstrap-deployer-vm.yml, when: True }
  - { playbook: ardana-ses-integration.yml, when: ses_enabled }
  - { playbook: shutdown-bm-nodes.yml, when: True }
  - { playbook: run-ardana-deploy-plays.yml, when: True }
  - { playbook: ardana-ses-rgw.yml, when: ses_enabled }
```
Though `include_playbook` might be needed. Unclear if this will work even then. Also it was said that `when` might not work well. Roles are probably better than using playbooks.

stefannica commented:
> looser coupling with Jenkins
> 
> I'd like for most of what is done in Jenkins jobs to work outside of a Jenkins server (say on a development machine). 
> <snipped>

@JanZerebecki I agree with the requirement of being able to run the pipeline automation logic without involving Jenkins, say from your local machine. I have removed everything that I thought was irrelevant from a development perspective from the Jenkins pipeline and used as few groovy specific commands as possible. Furthermore, almost every stage in the pipeline is now either invoking an outside bash script, or an ansible playbook, which can just as easily be replicated from a development environment.

The pipeline layouts will go through some important modifications in the next days, mostly to "merge" the stages that are now implemented differently for hardware and virtual environments. What I propose, after that is to revisit this proposal and attempt to generate the pipeline definitions from a common template. This task will also attempt to remove and externalize the remaining logic from the pipelines that would be required to implement a development workflow built around the same automation logic, and also to re-implement some of the remaining bash scripts as ansible playbooks (which can more easily be condensed into a comprehensive ansible based toolchain that can be used by developers - something that @rtamalin  is also interested in having).

stefannica commented:
rebased

stefannica commented on scripts/jenkins/ardana/README.md:
```diff
@@ -0,0 +1,43 @@
+  * setting the workspace
+  * setting the job name
+  * post build triggers
+  * downstream job triggers (part of the pipeline now)
```
@JanZerebecki currently, no, not in a way that includes all the features of using the pipeline specific plugin, e.g. the ability to pass environment variables upstream and to abort a downsteam job when the parent job/stage is aborted

stefannica commented on scripts/jenkins/ardana/README.md:
```diff
@@ -0,0 +1,43 @@
+* passing information from one stage to the next:
+  * use the shared workspace
+  * use ansible variables (host/group vars, extra vars) as much as possible
+  * use build environment variables (which can only be set by using `env.<variable-name>` in a pipeline script block)
```
@JanZerebecki unfortunately, that's not possible. The environment variables exported by shell scripts are not picked up by Jenkins. The only way to accomplish this is by using groovy `env.<variable-name>` variables

stefannica commented on scripts/jenkins/ardana/ansible/bin/heat_stack.sh:
```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env bash
```
@JanZerebecki I don't want to modify the current `openstack-ardana` and `openstack-ardana-heat` jobs yet. They will continue to coexist with the new pipeline jobs for a while, giving us the possibility to quickly switch back to the old jobs if things go sideways with the new ones. When the pipeline jobs are thoroughly tested, I will push another PR removing the openstack-ardana and openstack-ardana-heat jobs and their associated scripts, if any remain.

stefannica commented:
@JanZerebecki I updated the documentation to cover everything you pointed out, including your idea about generating the Jenkinsfile contents and the reached compromise. I also cleaned up remaining the Jenkinsfiles some more. Please review when you get a chance.
